donkeycar/parts/robocars_hat_ctrl.py

```diff
@@ -0,0 +1,345 @@
+        elif feature == 'record':
+            return self.AUX_FEATURE_RECORD
+        elif feature == 'pilot':
+            return self.AUX_FEATURE_PILOT
+        elif feature == 'throttle_exploration':
+            return self.AUX_FEATURE_THROTTLEEXP
+        elif feature == 'steering_exploration':
+            return self.AUX_FEATURE_STEERINGEXP
+        elif feature == 'output_steering_trim':
+            return self.AUX_FEATURE_OUTPUT_STEERING_TRIM
+        elif feature == 'output_steering_exp':
+            return self.AUX_FEATURE_OUTPUT_STEERING_EXP
+
+    def __init__(self, cfg):
+
+        self.cfg = cfg
+        self.inSteering = 0.0
+        self.inThrottle = 0.0
+        self.fixThrottle = 0.0
+        self.fixSteering = 0.0
+        self.fixOutputSteeringTrim = None
+        self.fixOutputSteering = None
+        self.inAux1 = 0.0
+        self.inAux2 = 0.0
+        self.lastAux1 = -1.0
+        self.lastAux2 = -1.0
+        self.recording=False
+        self.mode = 'user'
+        self.lastMode = self.mode
+        self.applyBrake = 0
+
+        if (self.cfg.ROBOCARSHAT_USE_AUTOCALIBRATION==True) :
+            self.inThrottleIdle = -1
+            self.inSteeringIdle = -1
+        else:
+            self.inThrottleIdle = 1500
+            self.inSteeringIdle = 1500
+
+        self.inSpeed = 0
+
+        #Aux feature
+        self.ch3Feature = self.AUX_FEATURE_NONE
+        self.ch4Feature = self.AUX_FEATURE_NONE
+
+        self.ch3Feature = self._map_aux_feature (self.cfg.ROBOCARSHAT_CH3_FEATURE)
+        self.ch4Feature = self._map_aux_feature (self.cfg.ROBOCARSHAT_CH4_FEATURE)
+        if (self.ch3Feature == self.AUX_FEATURE_OUTPUT_STEERING_TRIM) or (self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_TRIM):
+            self.fixOutputSteeringTrim = 1500
+        if (self.ch3Feature == self.AUX_FEATURE_OUTPUT_STEERING_EXP) or (self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_EXP):
+            self.fixOutputSteering = 1500
+
+        if self.cfg.ROBOCARSHAT_THROTTLE_DISCRET != None:
+            self.discretesThrottle = np.arange(0.0,1.0001,1.0/len(self.cfg.ROBOCARSHAT_THROTTLE_DISCRET))
+            mylogger.info("CtrlIn Discrete throttle thresholds set to {}".format(self.discretesThrottle))
+
+        self.sensor = RobocarsHat(self.cfg)
+        self.on = True
+
+        if self.cfg.ROBOCARSHAT_EMERGENCY_STOP == True:
+            self.emergemcyPort=9111
+            self.stop=False
+            self.server = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+            self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+            self.server.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+            fcntl.fcntl(self.server, fcntl.F_SETFL, os.O_NONBLOCK)
+            self.server.bind(("", self.emergemcyPort))
+            print("Listening emergency on port {}".format(self.emergemcyPort))
+
+    def map_range(self, x, X_min, X_max, Y_min, Y_max):
```

Comment:
I don't like the fact that you are copying this method again, I see that it is already a bit messed up in the `utils.py` module, but we don't need the same thing 3x. Best would be to define the algorithm once (like here) and then have two versions, one which casts to `int` and another that rounds, both of these just calling the base algorithm. Or add an additional argument that does the rounding / casting on the result, if needed.